### PR TITLE
feat[contracts]: Replace Lib_RingBuffer with a much simpler Lib_Buffer

### LINKED ISSUE #611: Remove Lib_RingBuffer
**Is your feature request related to a problem? Please describe.**
We're still using `Lib_RingBuffer` but we should really get rid of it. It adds unnecessary complexity + can be replaced by simply using a standard array. Contract here: https://github.com/ethereum-optimism/optimism/blob/master/packages/contracts/contracts/optimistic-ethereum/libraries/utils/Lib_RingBuffer.sol

**Describe the solution you'd like**
We should remove `Lib_RingBuffer` and replace its usage inside `OVM_ChainStorageContainer` with a simple array.

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Replaces `Lib_RingBuffer` with a vastly simplified `Lib_Buffer`. I find that this library is still pretty useful as an array that can be rolled back and overwritten (+ the extra data is nice!).

Closes #611.

### CODE REVIEW

**Comment by tynes** [comment]
Are there no unit tests for this? I see none of the existing tests break

**Comment by maurelian** [comment]
I agree this should have unit tests, but perhaps that can wait for the testing project for which [the charter](https://www.notion.so/optimismpbc/Improved-Contracts-Testing-fd8455e3eb0d460787435d795b8c4342#f89db2b8dc494bbcbc557cbef369944a) is currently being finalized, and includes testing the Lib_Buffer. 

I'm not sure if that means this can be merged, or if test tests should be written into this branch during the next sprint.

**Comment by maurelian** [comment]
FWIW, I spoke out of band with @smartcontracts this morning, and he intends to add tests to this PR. I'll move it to draft for now.

**Comment by snario** [comment]
This needs to re-target `regenesis/0.4.0`.

File: packages/contracts/contracts/optimistic-ethereum/libraries/utils/Lib_Buffer.sol [hunk 1]
```diff
@@ -0,0 +1,251 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >0.5.0 <0.8.0;
+
+/**
+ * @title Lib_Buffer
+ * @dev Essentially just a fancy array.
```
Comment (ben-chain) [inline packages/contracts/contracts/optimistic-ethereum/libraries/utils/Lib_Buffer.sol:6]:
```suggestion
 * @dev This library implements a bytes32 storage array with some additional gas-optimized functionality.  In particular, it encodes its length as a uint40, and tightly packs this with an overwritable "scratch pad" field into a single SSTORE.
```

File: packages/contracts/contracts/optimistic-ethereum/libraries/utils/Lib_Buffer.sol [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+     ***********/
+
+    struct Buffer {
+        bytes32 context;
+        mapping (uint256 => bytes32) buf;
+    }
+
+    struct BufferContext {
+        uint40 length;
+        bytes27 extraData;
```
Comment (ben-chain) [inline packages/contracts/contracts/optimistic-ethereum/libraries/utils/Lib_Buffer.sol:28]: Would love to see some comments here explaining the nature of `extraData`.

**Review by ben-chain — APPROVED** [review]
Left a few suggestions for improving codebase readability with comments, but approving in advance -- looks great.  Love that almost-negative diff where we still are adding tests!

File: packages/contracts/contracts/optimistic-ethereum/libraries/utils/Lib_Buffer.sol (hunk 2 above)
Comment (smartcontracts) [reply packages/contracts/contracts/optimistic-ethereum/libraries/utils/Lib_Buffer.sol]: Done.

**Review by elenadimitrova — APPROVED** [review]
Can you pls update PR description to include "Closes #611".

Test coverage of the `Lib_Buffer` is good and >80%

Notable gas savings:

`OVM_CanonicalTransactionChain`
`appendSequencerBatch` -40,256 gas
`enqueue` -14,197 gas

`OVM_L1CrossDomainMessenger`
`replayMessage` -14,592 gas
`sendMessage` -15,307 gas

`OVM_StateCommitmentChain`
`appendStateBatch` -19,000 gas
`deleteStateBatch`  -7,246 gas

**Pull request merged**